abdiff/core/calc_ab_metrics.py

```diff
@@ -0,0 +1,291 @@
+# ruff: noqa: S608
+
+import json
+import logging
+import os
+import time
+from pathlib import Path
+
+import duckdb
+import pandas as pd
+import pyarrow.dataset as ds
+from duckdb.duckdb import DuckDBPyConnection
+
+from abdiff.core.utils import update_or_create_run_json
+
+logger = logging.getLogger(__name__)
+
+NON_TIMDEX_FIELD_COLUMNS = ["timdex_record_id", "source"]
+
+
+def calc_ab_metrics(
+    run_directory: str,
+    diffs_dataset: str,
+) -> dict:
+
+    os.makedirs(Path(run_directory) / "metrics", exist_ok=True)
+
+    # build field diffs dataframe
+    field_matrix_parquet = create_record_diff_matrix_parquet(run_directory, diffs_dataset)
+
+    # calculate metrics data from sparse matrix
+    metrics_data = calculate_metrics_data(field_matrix_parquet)
+
+    # update run data with metrics
+    update_or_create_run_json(
+        run_directory=run_directory, new_data={"metrics": metrics_data}
+    )
+
+    return metrics_data
+
+
+def create_record_diff_matrix_parquet(
+    run_directory: str,
+    diffs_dataset: str,
+    batch_size: int = 1_000,
+) -> str:
+    """Create a boolean sparse matrix of modified fields for all records.
```

Comment:
Example of the output parquet file rows / columns:

```
source, timdex_record_id, dates, contributors, locations, ...
------------------------------------------------------------
alma, alma:123, 0, 0, 0, ...
alma, alma:345, 1, 0, 1, ...
dspace, dspace:789, 0, 0, 1, ...
```

Despite the potentially high number of rows, this data will be quite small in memory.